relays/ethereum/src/message_lane_loop.rs

```diff
@@ -0,0 +1,591 @@
+
+	/// Get nonce of latest message, which receival has been confirmed.
+	async fn latest_received_nonce(
+		&self,
+		id: TargetHeaderIdOf<P>,
+	) -> Result<(TargetHeaderIdOf<P>, P::MessageNonce), Self::Error>;
+
+	/// Submit messages proof.
+	async fn submit_messages_proof(
+		&self,
+		generated_at_header: SourceHeaderIdOf<P>,
+		nonces: RangeInclusive<P::MessageNonce>,
+		proof: P::MessagesProof,
+	) -> Result<RangeInclusive<P::MessageNonce>, Self::Error>;
+}
+
+/// State of the client.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct ClientState<SelfHeaderId, PeerHeaderId> {
+	/// Best header id of this chain.
+	pub best_self: SelfHeaderId,
+	/// Best header id of the peer chain.
+	pub best_peer: PeerHeaderId,
+}
+
+/// State of source client in one-way message lane.
+pub type SourceClientState<P> = ClientState<SourceHeaderIdOf<P>, TargetHeaderIdOf<P>>;
+
+/// State of target client in one-way message lane.
+pub type TargetClientState<P> = ClientState<TargetHeaderIdOf<P>, SourceHeaderIdOf<P>>;
+
+/// Both clients state.
+#[derive(Debug, Default)]
+pub struct ClientsState<P: MessageLane> {
+	/// Source client state.
+	pub source: Option<SourceClientState<P>>,
+	/// Target client state.
+	pub target: Option<TargetClientState<P>>,
+}
+
+/// Run message lane service loop.
+pub fn run<P: MessageLane>(
+	mut source_client: impl SourceClient<P>,
+	source_tick: Duration,
+	mut target_client: impl TargetClient<P>,
+	target_tick: Duration,
+	reconnect_delay: Duration,
+	stall_timeout: Duration,
+	exit_signal: impl Future<Output = ()>,
+) {
+	let mut local_pool = futures::executor::LocalPool::new();
+	let exit_signal = exit_signal.shared();
+
+	local_pool.run_until(async move {
+		loop {
+			let result = run_until_connection_lost(
+				source_client.clone(),
+				source_tick,
+				target_client.clone(),
+				target_tick,
+				stall_timeout,
+				exit_signal.clone(),
+			)
+			.await;
+
+			match result {
+				Ok(()) => break,
+				Err(failed_client) => {
+					async_std::task::sleep(reconnect_delay).await;
+					if failed_client == FailedClient::Both || failed_client == FailedClient::Source {
+						source_client = source_client.reconnect();
+					}
+					if failed_client == FailedClient::Both || failed_client == FailedClient::Target {
+						target_client = target_client.reconnect();
+					}
+				}
+			}
+
+			log::debug!(
+				target: "bridge",
+				"Restarting lane {} -> {}",
+				P::SOURCE_NAME,
+				P::TARGET_NAME,
+			);
+		}
+	});
+}
+
+/// Run one-way message delivery loop until connection with target or source node is lost, or exit signal is received.
+async fn run_until_connection_lost<P: MessageLane, SC: SourceClient<P>, TC: TargetClient<P>>(
+	source_client: SC,
+	source_tick: Duration,
+	target_client: TC,
+	target_tick: Duration,
+	stall_timeout: Duration,
+	exit_signal: impl Future<Output = ()>,
+) -> Result<(), FailedClient> {
+	let mut source_retry_backoff = retry_backoff();
+	let mut source_client_is_online = false;
```

Comment:
I'm still not sold on the `online/offline` concept, but since it's prevalent in other sync loops already, I can accept it.